### ISSUE TITLE
Reimplementing core functions

### ISSUE THREAD

**Issue by Misiur** [pawn-lang/YSI-Includes#193]
I dug out an old project of mine, [`ut_mock_players`](http://forum.sa-mp.com/showthread.php?t=490767) and I want to add it to YSI. It had a wrong assumption that all players should be "connected" for tests, while it is not true, as often we need testcases for not connected players. So, now instead of nifty macros, I have to reimplement samp's own functions. For example: [SetPlayerName](http://wiki.sa-mp.com/wiki/SetPlayerName) - it turns out it has a return code, and it is not meaningless. 
```pawn
stock Y_MOCK_SetPlayerName(playerid, name[]) {
    if (!IsPlayerConnected(playerid)) {
        return 0;
    }

    if (isnull(name)) {
        return CRASH_VALUE;
    }

    if (!isnull(MockPlayer[playerid][E_USERNAME]) && !strcmp(MockPlayer[playerid][E_USERNAME], name)) {
        return 0;
    }

    if (strlen(name) > MAX_PLAYER_NAME) {
        return -1;
    }

    // Regex would be easier
    for (new i = 0; i != MAX_PLAYER_NAME + 1; ++i) {
        if (!
            ('a' < name[i] < 'Z'
            || '0' < name[i] < '9'
            || name[i] != '['
            || name[i] != ']'
            || name[i] != '('
            || name[i] != ')'
            || name[i] != '$'
            || name[i] != '@'
            || name[i] != '['
            || name[i] != '.'
            || name[i] != '_'
            || name[i] != '='
            )
        ) {
            return -1;
        }
    }

    strcpy(MockPlayer[playerid][E_USERNAME], name);

    return 1;  
}

#define SetPlayerName Y_MOCK_SetPlayerName
```

As you can see I can't use `_ALS_` anymore, because the underlaying native function return value overrides my code. 

So long story short - is there any option to execute all hooked functions, but not the native one?

**Comment by Y-Less** [comment]
I'm not sure what you mean by not being able to use ALS.

**Comment by Y-Less** [comment]
I'm not sure what you mean by not being able to use ALS.

**Comment by Y-Less** [comment]
But if the hooks are included after a_samp, but before everything else, then yes they will be called instead.

**Comment by Misiur** [comment]
Gah, I didn't call original function in my als tests and I mistakenly assumed that only last defined hook is called. This is much better!

**Comment by Southclaws** [comment]
Completely forgot about that library, glad to see it joining YSI! I must get back into the habit of writing unit tests in Pawn again...

**Comment by Y-Less** [comment]
> I should write more tests

- Every coder ever.

**Comment by Y-Less** [comment]
I also forgot about that library, I'd say I'd never heard of it, but my multiple posts in the topic prove otherwise...

Also:

#157

**Comment by Misiur** [comment]
Well, it was not even finished, no reason to remember obscure, unused, unfinished lib.
Before I sink my time into it let me explain my usecase and tell me your opinion if it makes any sense to continue. Say, there's a command for single teleport for every even user

```pawn
new Foo[MAX_PLAYERS] = { -1, 1, ... };

YCMD:bar(playerid, params[], help) 
{
	if (Foo[playerid] == -1) {
        	Foo[playerid] = 1;
		SetPlayerPos(playerid, 100.0, 20.0, 40.0);
	}

	return 1;
}
```

Now I want to be able to write a test for it without the need for firing up samp

```pawn
TestInit:SingleTeleport()
{
	CreatePlayer(0);
	CreatePlayer(3);
}


Test:SingleTeleport()
{
	new Float:X, Float:Y, Float:Z;
	GetPlayerPos(0, X, Y, Z);
	ASSERT_TRUE((X == 0.0 && Y == 0.0 && Z == 0.0));
	CallLocalFunction("OnPlayerCommandText", "is", 0, "/bar");
	GetPlayerPos(0, X, Y, Z);
	ASSERT_TRUE((X == 100.0 && Y == 20.0 && Z == 40.0));
	ASSERT(Foo[0] == 1);

	GetPlayerPos(3, X, Y, Z);
	ASSERT_TRUE((X == 0.0 && Y == 0.0 && Z == 0.0));
	CallLocalFunction("OnPlayerCommandText", "is", 3, "/bar");
	GetPlayerPos(3, X, Y, Z);
	ASSERT_TRUE((X == 0.0 && Y == 0.0 && Z == 0.0));
}

TestClose:SingleTeleport()
{
	DeletePlayer(0);
	DeletePlayer(3);
}
```

Is it even possible to trigger command using CLF? Also, would y_testing be able to handle async tests (say, Interpolate* functions testing)?

**Comment by Misiur** [comment]
Alright, it is possible with my modified IsPlayerConnected. As for async, well, timers? Don't really know, don't want to clog the main thread

**Comment by Y-Less** [comment]
y_testing doesn't do later responses amazingly.  It can handle them sort of, but currently only for players, in that it can request a "did this work?" response from a player.  If you want more deferred responses, I'm not sure of the best way to do that right now sorry.

**Comment by Southclaws** [comment]
async stuff was one of the main things I wanted to implement when I opened #157 - I've been working with quite a few interesting testing libraries and techniques in Go recently so I might take another look at it once I have the time.
